File: editors/jetbrains/src/test/kotlin/com/github/btakita/agentdoc/PeerReplicaPullTest.kt
package com.github.btakita.agentdoc

import java.nio.file.Files
import java.nio.file.Paths
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * `#ctrlkillreregister` Tier 3 — the editor-side pull decision.
 *
 * These assert the two distinctions the caller's correctness hangs on: "could not
 * ask" is not "nothing to do", and another peer's stranded registration is not this
 * editor's to rebuild.
 */
class PeerReplicaPullTest {
    private fun registration(pid: Long, path: String): String =
        """{"document_hash":"h-$pid","pid":$pid,"path":"$path","editor_id":"jetbrains-$pid",""" +
            """"editor_kind":"jetbrains","editor_version":"0.2.0","capabilities":[],"timestamp_ms":1}"""

    @Test
    fun `an up-to-date editor is told to rebuild nothing`() {
        assertEquals(emptyList<String>(), PeerReplicaPull.rebuildPaths("[]", 42L))
    }

    @Test
    fun `an unanswerable pull is distinct from an empty answer`() {
        // Null must NOT collapse into "nothing to do": the caller falls back to the
        // controller's compatibility fan-out on null, and doing nothing instead would
        // leave the editor stranded exactly when the pull is unavailable.
        assertNull("a missing ABI/controller cannot mean up to date", PeerReplicaPull.rebuildPaths(null, 42L))
        assertNull(PeerReplicaPull.rebuildPaths("", 42L))
        assertNull(PeerReplicaPull.rebuildPaths("   ", 42L))
        assertNull("malformed JSON is not an empty answer", PeerReplicaPull.rebuildPaths("{not json", 42L))
        assertNull("a non-array answer is not an empty answer", PeerReplicaPull.rebuildPaths("""{"pid":42}""", 42L))

        assertNotNull("a real empty answer stays empty", PeerReplicaPull.rebuildPaths("[]", 42L))
    }

    @Test
    fun `only this editor's own stranded registrations are rebuilt`() {
        val json = "[${registration(42L, "/proj/mine.md")},${registration(7L, "/proj/theirs.md")}]"

        assertEquals(
            "rebuilding another editor's document would publish this buffer's text over theirs",
            listOf("/proj/mine.md"),
            PeerReplicaPull.rebuildPaths(json, 42L),
        )
        assertEquals(listOf("/proj/theirs.md"), PeerReplicaPull.rebuildPaths(json, 7L))
        assertEquals(emptyList<String>(), PeerReplicaPull.rebuildPaths(json, 9999L))
    }

    @Test
    fun `duplicate and malformed entries do not produce duplicate or empty rebuild targets`() {
        val json = "[" +
            registration(42L, "/proj/a.md") + "," +
            registration(42L, "/proj/a.md") + "," +
            """{"pid":42}""" + "," +
            """{"path":"/proj/b.md"}""" + "," +
            """{"pid":42,"path":""}""" + "," +
            "\"not-an-object\"" +
            "]"

        assertEquals(
            "one re-register per stranded path; entries without a pid+path are skipped",
            listOf("/proj/a.md"),
            PeerReplicaPull.rebuildPaths(json, 42L),
        )
    }

    /**
     * The pull must be what runs at startup. A blind refresh of every open document
     * re-registers healthy replicas — dropping and rebuilding a live CRDT baseline —
     * and still misses a stranded registration whose document is not open in a tab.
     */
    @Test
    fun `startup pulls missing replicas instead of blindly refreshing every open document`() {
        val source = Files.readString(
            Paths.get("src/main/kotlin/com/github/btakita/agentdoc/PluginLifecycleListener.kt")
                .takeIf { Files.exists(it) }
                ?: Paths.get("editors/jetbrains/src/main/kotlin/com/github/btakita/agentdoc/PluginLifecycleListener.kt"),
        )

        assertTrue(
            source.contains("CrdtReplicaManager.pullMissingReplicas(project, \"plugin-startup\")"),
        )
        assertTrue(
            "the blind sweep must not remain a startup call site",
            !source.contains("forceRefreshOpenDocumentReplicas(project, \"plugin-startup\")"),
        )
    }

    /**
     * The pull is also the reconnect path: one document noticing controller transport
     * loss means every registration this editor holds is stranded, including ones
     * nothing is currently draining.
     */
    @Test
    fun `controller transport recovery pulls for the whole editor, coalesced`() {
        val source = Files.readString(
            Paths.get("src/main/kotlin/com/github/btakita/agentdoc/CrdtReplicaManager.kt")
                .takeIf { Files.exists(it) }
                ?: Paths.get("editors/jetbrains/src/main/kotlin/com/github/btakita/agentdoc/CrdtReplicaManager.kt"),
        )

        val recovery = source.substringAfter("private fun refreshReplicaAfterTransportLoss")
            .substringBefore("private fun shouldAttemptRegister")
        assertTrue(
            "transport recovery must ask about the rest of this editor's registrations",
            recovery.contains("pullMissingReplicas(project, \"controller-transport-recovered\")"),
        )

        val pull = source.substringAfter("fun pullMissingReplicas(")
            .substringBefore("fun forceRefreshOpenDocumentReplicas")
        assertTrue(
            "a dead controller strands every document at once; one pull answers for all",
            pull.contains("beginPeerReplicaPull()"),
        )
        assertTrue(
            "held must stay empty: stale local forwarders would suppress the repair",
            pull.contains("emptyList()"),
        )
        assertTrue(
            "an unaskable pull must fall back, not silently do nothing",
            pull.contains("forceRefreshOpenDocumentReplicas(project, \"\$reason-pull-unavailable\")"),
        )
    }

    /**
     * The capability token is the controller's retirement condition for the Tier 1
     * fan-out. Advertising it without calling the pull would silence the push while
     * nothing repaired — strictly worse than either tier alone.
     */
    @Test
    fun `the advertised pull capability matches the code that actually pulls`() {
        val capabilities = Files.readString(
            Paths.get("src/main/kotlin/com/github/btakita/agentdoc/TypingTracker.kt")
                .takeIf { Files.exists(it) }
                ?: Paths.get("editors/jetbrains/src/main/kotlin/com/github/btakita/agentdoc/TypingTracker.kt"),
        )
        assertTrue(capabilities.contains("\"peer_replica_pull_v1\""))
        assertTrue(capabilities.contains("add(PEER_REPLICA_PULL_CAPABILITY)"))
        assertTrue(
            "the token must be advertised in the registration payload",
            EDITOR_CAPABILITIES.split(",").contains("peer_replica_pull_v1"),
        )

        val pullSource = Files.readString(
            Paths.get("src/main/kotlin/com/github/btakita/agentdoc/PeerReplicaPull.kt")
                .takeIf { Files.exists(it) }
                ?: Paths.get("editors/jetbrains/src/main/kotlin/com/github/btakita/agentdoc/PeerReplicaPull.kt"),
        )
        assertTrue(
            "the capability claims this exact FFI call",
            pullSource.contains("agent_doc_peer_replicas_missing("),
        )
    }
}

File: editors/vscode/src/peerReplicaPull.test.ts
import { describe, it } from 'node:test';
import assert from 'node:assert';
import { readFileSync } from 'node:fs';
import { peerReplicaRebuildPaths } from './peerReplicaPull.js';
import { CrdtReplicaManager, type ReplicaNode, type ReplicaTransport } from './crdtReplica.js';
import { EDITOR_CAPABILITY_LIST } from './native.js';

function registration(pid: number, path: string): string {
    return JSON.stringify({
        document_hash: `h-${pid}`,
        pid,
        path,
        editor_id: `vscode-${pid}`,
        editor_kind: 'vscode',
        editor_version: '0.2.57',
        capabilities: [],
        timestamp_ms: 1,
    });
}

/**
 * `#ctrlkillreregister` Tier 3 — the editor-side pull decision.
 *
 * These assert the two distinctions the caller's correctness hangs on: "could not
 * ask" is not "nothing to do", and another peer's stranded registration is not this
 * editor's to rebuild.
 */
describe('peerReplicaRebuildPaths', () => {
    it('reports nothing to rebuild for an up-to-date editor', () => {
        assert.deepStrictEqual(peerReplicaRebuildPaths('[]', 42), []);
    });

    it('keeps an unanswerable pull distinct from an empty answer', () => {
        // Null must NOT collapse into "nothing to do": the caller relies on the
        // controller's compatibility fan-out when the pull cannot be asked, and
        // silently doing nothing would leave the editor stranded exactly then.
        assert.strictEqual(peerReplicaRebuildPaths(null, 42), null);
        assert.strictEqual(peerReplicaRebuildPaths('', 42), null);
        assert.strictEqual(peerReplicaRebuildPaths('   ', 42), null);
        assert.strictEqual(peerReplicaRebuildPaths('{not json', 42), null);
        assert.strictEqual(peerReplicaRebuildPaths('{"pid":42}', 42), null);

        assert.notStrictEqual(peerReplicaRebuildPaths('[]', 42), null);
    });

    it('rebuilds only this editor process’s own stranded registrations', () => {
        const json = `[${registration(42, '/proj/mine.md')},${registration(7, '/proj/theirs.md')}]`;

        assert.deepStrictEqual(
            peerReplicaRebuildPaths(json, 42),
            ['/proj/mine.md'],
            'rebuilding another editor’s document would publish this buffer’s text over theirs',
        );
        assert.deepStrictEqual(peerReplicaRebuildPaths(json, 7), ['/proj/theirs.md']);
        assert.deepStrictEqual(peerReplicaRebuildPaths(json, 9999), []);
    });

    it('skips duplicate and malformed entries', () => {
        const json = `[${[
            registration(42, '/proj/a.md'),
            registration(42, '/proj/a.md'),
            '{"pid":42}',
            '{"path":"/proj/b.md"}',
            '{"pid":42,"path":""}',
            '"not-an-object"',
            'null',
        ].join(',')}]`;

        assert.deepStrictEqual(peerReplicaRebuildPaths(json, 42), ['/proj/a.md']);
    });
});

class SilentNode implements ReplicaNode {
    open(): boolean {
        return true;
    }
    applyLocal(): boolean {
        return true;
    }
    applyUpdate(): boolean {
        return true;
    }
    encodeState(): Uint8Array {
        return Buffer.from([1]);
    }
    text(): string {
        return 'base';
    }
    close(): void {}
}

class SilentTransport implements ReplicaTransport {
    registered: string[] = [];
    async register(filePath: string) {
        this.registered.push(filePath);
        return { clientId: 42 };
    }
    async broadcastUpdate(): Promise<void> {}
    async pullUpdates() {
        return [];
    }
    async deregister(): Promise<void> {}
}

function manager(
    peerReplicasMissing: (pid: number, held: readonly string[]) => string | null,
    transport = new SilentTransport(),
): { manager: CrdtReplicaManager; transport: SilentTransport; asked: string[][] } {
    const asked: string[][] = [];
    return {
        transport,
        asked,
        manager: new CrdtReplicaManager({
            projectRoot: '/work',
            identity: 'vscode-test',
            pid: 42,
            transport,
            nodeFactory: () => new SilentNode(),
            listDocuments: () => [],
            currentText: () => 'base',
            applyText: async () => true,
            peerReplicasMissing: (pid, held) => {
                asked.push([...held]);
                return peerReplicasMissing(pid, held);
            },
        }),
    };
}

describe('CrdtReplicaManager.pullMissingReplicas', () => {
    it('re-registers exactly the registrations the controller says it cannot serve', async () => {
        const { manager: mgr, transport } = manager(() =>
            `[${registration(42, '/work/stranded.md')}]`,
        );

        mgr.seedDocument('/work/stranded.md', 'base');
        await mgr.pullMissingReplicas('test');

        assert.deepStrictEqual(transport.registered, ['/work/stranded.md']);
    });

    it('rebuilds nothing when the editor is up to date', async () => {
        const { manager: mgr, transport } = manager(() => '[]');

        await mgr.pullMissingReplicas('test');

        assert.deepStrictEqual(
            transport.registered,
            [],
            'a healthy replica must never be dropped and rebuilt',
        );
    });

    it('asks with an empty held set so stale local forwarders cannot suppress the repair', async () => {
        // After a controller kill this editor’s forwarders still look live here, so
        // passing them as `held` would suppress precisely the documents that need
        // repair. The controller subtracts what its own hub can serve instead.
        const { manager: mgr, asked } = manager(() => '[]');

        mgr.seedDocument('/work/looks-live.md', 'base');
        await mgr.attachDocument('/work/looks-live.md');
        await mgr.pullMissingReplicas('test');

        assert.deepStrictEqual(asked, [[]]);
    });

    it('coalesces concurrent pulls: one controller death is one question', async () => {
        let calls = 0;
        const { manager: mgr } = manager(() => {
            calls += 1;
            return '[]';
        });

        await mgr.pullMissingReplicas('transport-loss-doc-a');
        await mgr.pullMissingReplicas('transport-loss-doc-b');
        await mgr.pullMissingReplicas('transport-loss-doc-c');

        assert.strictEqual(calls, 1, 'every open document reports the same controller death');
    });

    it('does not force-refresh healthy replicas when the pull cannot be asked', async () => {
        const { manager: mgr, transport } = manager(() => null);

        mgr.seedDocument('/work/plan.md', 'base');
        await mgr.pullMissingReplicas('test');

        assert.deepStrictEqual(
            transport.registered,
            [],
            'null means "could not ask"; the existing attach/retry paths remain the fallback',
        );
    });
});

describe('peer replica pull capability', () => {
    /**
     * The capability token is the controller’s retirement condition for the Tier 1
     * fan-out. Advertising it without calling the pull would silence the push while
     * nothing repaired — strictly worse than either tier alone.
     */
    it('is advertised only alongside the code that actually pulls', () => {
        assert.ok(EDITOR_CAPABILITY_LIST.includes('peer_replica_pull_v1'));

        const nativeSource = readFileSync(new URL('../src/native.ts', import.meta.url), 'utf8');
        assert.ok(nativeSource.includes("'agent_doc_peer_replicas_missing'"));

        const replicaSource = readFileSync(new URL('../src/crdtReplica.ts', import.meta.url), 'utf8');
        assert.ok(
            replicaSource.includes("this.pullMissingReplicas('controller-transport-recovered')"),
            'reconnect must pull for the whole editor, not just the file that noticed',
        );
        assert.ok(
            replicaSource.includes("this.pullMissingReplicas('activation')"),
            'activation must pull as a safety net after the normal attach pass',
        );
    });
});
